From 316aab9eb3f80084015b7d24bfaebc007a269f83 Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Thu, 3 Nov 2022 15:23:39 +0800
Subject: [PATCH 190/595] [Huawei] cgroup: Introduce CONFIG_CGROUP_V1_WRITEBACK

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187773

-------------------------------

Isolate cgroup v1 writeback feature with CONFIG_CGROUP_V1_WRITEBACK
instead of CONFIG_CGROUP_WRITEBACK.

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 arch/arm64/configs/openeuler_defconfig |  1 +
 arch/x86/configs/openeuler_defconfig   |  1 +
 block/blk-cgroup.c                     |  2 ++
 include/linux/backing-dev.h            | 30 +++++++++++++++++---------
 include/linux/blk-cgroup.h             |  2 +-
 include/linux/cgroup.h                 |  2 +-
 include/linux/memcontrol.h             |  4 ++--
 init/Kconfig                           |  5 +++++
 kernel/cgroup/cgroup.c                 |  9 +-------
 mm/backing-dev.c                       | 10 ++++++++-
 mm/memcontrol.c                        |  6 ++----
 11 files changed, 45 insertions(+), 27 deletions(-)

diff --git a/arch/arm64/configs/openeuler_defconfig b/arch/arm64/configs/openeuler_defconfig
index 505f822d7b55..4a36cb8b357e 100644
--- a/arch/arm64/configs/openeuler_defconfig
+++ b/arch/arm64/configs/openeuler_defconfig
@@ -136,6 +136,7 @@ CONFIG_MEMCG_KMEM=y
 CONFIG_MEMCG_MEMFS_INFO=y
 CONFIG_BLK_CGROUP=y
 CONFIG_CGROUP_WRITEBACK=y
+CONFIG_CGROUP_V1_WRITEBACK=y
 CONFIG_CGROUP_SCHED=y
 CONFIG_FAIR_GROUP_SCHED=y
 CONFIG_CFS_BANDWIDTH=y
diff --git a/arch/x86/configs/openeuler_defconfig b/arch/x86/configs/openeuler_defconfig
index 57d0871a4cdf..67478ceeea0e 100644
--- a/arch/x86/configs/openeuler_defconfig
+++ b/arch/x86/configs/openeuler_defconfig
@@ -155,6 +155,7 @@ CONFIG_MEMCG_KMEM=y
 CONFIG_MEMCG_MEMFS_INFO=y
 CONFIG_BLK_CGROUP=y
 CONFIG_CGROUP_WRITEBACK=y
+CONFIG_CGROUP_V1_WRITEBACK=y
 CONFIG_CGROUP_SCHED=y
 CONFIG_QOS_SCHED=y
 CONFIG_QOS_SCHED_SMT_EXPELLER=y
diff --git a/block/blk-cgroup.c b/block/blk-cgroup.c
index dd1193494016..bdfafb51ab58 100644
--- a/block/blk-cgroup.c
+++ b/block/blk-cgroup.c
@@ -1146,6 +1146,8 @@ blkcg_css_alloc(struct cgroup_subsys_state *parent_css)
 	INIT_HLIST_HEAD(&blkcg->blkg_list);
 #ifdef CONFIG_CGROUP_WRITEBACK
 	INIT_LIST_HEAD(&blkcg->cgwb_list);
+#endif
+#ifdef CONFIG_CGROUP_V1_WRITEBACK
 	INIT_LIST_HEAD(&blkcg->memcg_list);
 #endif
 	list_add_tail(&blkcg->all_blkcgs_node, &all_blkcgs);
diff --git a/include/linux/backing-dev.h b/include/linux/backing-dev.h
index ddbb0e6ff40f..e598848edcee 100644
--- a/include/linux/backing-dev.h
+++ b/include/linux/backing-dev.h
@@ -167,6 +167,26 @@ static inline int bdi_sched_wait(void *word)
 	return 0;
 }
 
+#ifdef CONFIG_CGROUP_V1_WRITEBACK
+void wb_kill_memcg_blkcg(struct cgroup_subsys_state *css);
+void wb_attach_memcg_to_blkcg(struct cgroup_subsys_state *memcg_css,
+			      struct cgroup_subsys_state *blkcg_css);
+bool cgroup1_writeback_enabled(void);
+#else
+static inline void wb_kill_memcg_blkcg(struct cgroup_subsys_state *css)
+{
+}
+static inline void
+wb_attach_memcg_to_blkcg(struct cgroup_subsys_state *memcg_css,
+			 struct cgroup_subsys_state *blkcg_css)
+{
+}
+static inline bool cgroup1_writeback_enabled(void)
+{
+	return false;
+}
+#endif	/* CONFIG_CGROUP_V1_WRITEBACK */
+
 #ifdef CONFIG_CGROUP_WRITEBACK
 
 struct bdi_writeback *wb_get_lookup(struct backing_dev_info *bdi,
@@ -178,11 +198,6 @@ void wb_memcg_offline(struct mem_cgroup *memcg);
 void wb_blkcg_offline(struct blkcg *blkcg);
 int inode_congested(struct inode *inode, int cong_bits);
 
-void wb_kill_memcg_blkcg(struct cgroup_subsys_state *css);
-void wb_attach_memcg_to_blkcg(struct cgroup_subsys_state *memcg_css,
-			      struct cgroup_subsys_state *blkcg_css);
-bool cgroup1_writeback_enabled(void);
-
 /**
  * inode_cgwb_enabled - test whether cgroup writeback is enabled on an inode
  * @inode: inode of interest
@@ -396,11 +411,6 @@ static inline int inode_congested(struct inode *inode, int cong_bits)
 	return wb_congested(&inode_to_bdi(inode)->wb, cong_bits);
 }
 
-static inline bool cgroup1_writeback_enabled(void)
-{
-	return false;
-}
-
 #endif	/* CONFIG_CGROUP_WRITEBACK */
 
 static inline int inode_read_congested(struct inode *inode)
diff --git a/include/linux/blk-cgroup.h b/include/linux/blk-cgroup.h
index 86a10c5acf9b..1fe9574b4a77 100644
--- a/include/linux/blk-cgroup.h
+++ b/include/linux/blk-cgroup.h
@@ -59,7 +59,7 @@ struct blkcg {
 #ifdef CONFIG_CGROUP_WRITEBACK
 	struct list_head		cgwb_list;
 #endif
-#if defined(CONFIG_CGROUP_WRITEBACK) && !defined(__GENKSYMS__)
+#if defined(CONFIG_CGROUP_V1_WRITEBACK) && !defined(__GENKSYMS__)
 	struct list_head		memcg_list;
 #else
 	KABI_RESERVE(1)
diff --git a/include/linux/cgroup.h b/include/linux/cgroup.h
index 744a15c536d8..8f43e5b5b203 100644
--- a/include/linux/cgroup.h
+++ b/include/linux/cgroup.h
@@ -696,7 +696,7 @@ static inline void cgroup_kthread_ready(void)
 }
 
 void cgroup_path_from_kernfs_id(u64 id, char *buf, size_t buflen);
-#ifdef CONFIG_CGROUP_WRITEBACK
+#ifdef CONFIG_CGROUP_V1_WRITEBACK
 struct cgroup *cgroup1_get_from_id(struct cgroup_root *root, u64 id);
 #endif
 #else /* !CONFIG_CGROUPS */
diff --git a/include/linux/memcontrol.h b/include/linux/memcontrol.h
index 27beafcb5fb6..bddf3c28a2fe 100644
--- a/include/linux/memcontrol.h
+++ b/include/linux/memcontrol.h
@@ -375,12 +375,12 @@ struct mem_cgroup {
 #ifdef CONFIG_DYNAMIC_HUGETLB
 	struct dhugetlb_pool *hpool;
 #endif
-#ifdef CONFIG_CGROUP_WRITEBACK
+#ifdef CONFIG_CGROUP_V1_WRITEBACK
 	KABI_USE(1, struct cgroup_subsys_state *wb_blk_css)
 #else
 	KABI_RESERVE(1)
 #endif
-#if defined(CONFIG_CGROUP_WRITEBACK) && !defined(__GENKSYMS__)
+#if defined(CONFIG_CGROUP_V1_WRITEBACK) && !defined(__GENKSYMS__)
 	struct list_head memcg_node;
 #else
 	KABI_RESERVE(2)
diff --git a/init/Kconfig b/init/Kconfig
index 9f44f636d930..c4cb83e2ce80 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -982,6 +982,11 @@ config CGROUP_WRITEBACK
 	depends on MEMCG && BLK_CGROUP
 	default y
 
+config CGROUP_V1_WRITEBACK
+	bool "Support Cgroup Writeback On Cgroupv1"
+	depends on CGROUP_WRITEBACK
+	default n
+
 menuconfig CGROUP_SCHED
 	bool "CPU controller"
 	default n
diff --git a/kernel/cgroup/cgroup.c b/kernel/cgroup/cgroup.c
index 055cf43458f5..a5dd82dd8b2c 100644
--- a/kernel/cgroup/cgroup.c
+++ b/kernel/cgroup/cgroup.c
@@ -59,10 +59,7 @@
 #include <linux/psi.h>
 #include <linux/dynamic_hugetlb.h>
 #include <net/sock.h>
-
-#ifdef CONFIG_CGROUP_WRITEBACK
 #include <linux/backing-dev.h>
-#endif
 
 #define CREATE_TRACE_POINTS
 #include <trace/events/cgroup.h>
@@ -5370,9 +5367,7 @@ static struct cgroup_subsys_state *css_create(struct cgroup *cgrp,
 	list_del_rcu(&css->sibling);
 err_free_css:
 	list_del_rcu(&css->rstat_css_node);
-#ifdef CONFIG_CGROUP_WRITEBACK
 	wb_kill_memcg_blkcg(css);
-#endif
 	INIT_RCU_WORK(&css->destroy_rwork, css_free_rwork_fn);
 	queue_rcu_work(cgroup_destroy_wq, &css->destroy_rwork);
 	return ERR_PTR(err);
@@ -5643,9 +5638,7 @@ static void kill_css(struct cgroup_subsys_state *css)
 	 */
 	css_get(css);
 
-#ifdef CONFIG_CGROUP_WRITEBACK
 	wb_kill_memcg_blkcg(css);
-#endif
 	/*
 	 * cgroup core guarantees that, by the time ->css_offline() is
 	 * invoked, no new css reference will be given out via
@@ -6090,7 +6083,7 @@ int proc_cgroup_show(struct seq_file *m, struct pid_namespace *ns,
 	return retval;
 }
 
-#ifdef CONFIG_CGROUP_WRITEBACK
+#ifdef CONFIG_CGROUP_V1_WRITEBACK
 struct cgroup *cgroup1_get_from_id(struct cgroup_root *root, u64 id)
 {
 	struct kernfs_node *kn;
diff --git a/mm/backing-dev.c b/mm/backing-dev.c
index 88287c35263d..707bf409f570 100644
--- a/mm/backing-dev.c
+++ b/mm/backing-dev.c
@@ -424,6 +424,7 @@ static void cgwb_remove_from_bdi_list(struct bdi_writeback *wb)
 	spin_unlock_irq(&cgwb_lock);
 }
 
+#ifdef CONFIG_CGROUP_V1_WRITEBACK
 static struct cgroup_subsys_state *cgwbv1_get_blkcss(struct mem_cgroup *memcg)
 {
 	struct cgroup_subsys_state *blkcg_css;
@@ -438,6 +439,13 @@ static struct cgroup_subsys_state *cgwbv1_get_blkcss(struct mem_cgroup *memcg)
 
 	return blkcg_css;
 }
+#else
+static inline struct cgroup_subsys_state *
+cgwbv1_get_blkcss(struct mem_cgroup *memcg)
+{
+	return NULL;
+}
+#endif
 
 static int cgwb_create(struct backing_dev_info *bdi,
 		       struct cgroup_subsys_state *memcg_css, gfp_t gfp)
@@ -1046,7 +1054,7 @@ long wait_iff_congested(int sync, long timeout)
 }
 EXPORT_SYMBOL(wait_iff_congested);
 
-#ifdef CONFIG_CGROUP_WRITEBACK
+#ifdef CONFIG_CGROUP_V1_WRITEBACK
 
 #include "../kernel/cgroup/cgroup-internal.h"
 
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 0a15761ba496..df62715b521b 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -5196,7 +5196,7 @@ static ssize_t memory_reclaim(struct kernfs_open_file *of, char *buf,
 	return nbytes;
 }
 
-#ifdef CONFIG_CGROUP_WRITEBACK
+#ifdef CONFIG_CGROUP_V1_WRITEBACK
 #include "../kernel/cgroup/cgroup-internal.h"
 
 static int wb_blkio_show(struct seq_file *m, void *v)
@@ -5463,7 +5463,7 @@ static struct cftype mem_cgroup_legacy_files[] = {
 		.name = "reclaim",
 		.write = memory_reclaim,
 	},
-#ifdef CONFIG_CGROUP_WRITEBACK
+#ifdef CONFIG_CGROUP_V1_WRITEBACK
 	{
 		.name = "wb_blkio_ino",
 		.flags = CFTYPE_NOT_ON_ROOT,
@@ -5731,9 +5731,7 @@ mem_cgroup_css_alloc(struct cgroup_subsys_state *parent_css)
 	if (cgroup_subsys_on_dfl(memory_cgrp_subsys) && !cgroup_memory_nosocket)
 		static_branch_inc(&memcg_sockets_enabled_key);
 
-#ifdef CONFIG_CGROUP_WRITEBACK
 	wb_attach_memcg_to_blkcg(&memcg->css, blkcg_root_css);
-#endif
 	return &memcg->css;
 fail:
 	mem_cgroup_id_remove(memcg);
-- 
2.17.1

